File: pe/nonvisualstudio/clojure/Implementation/Problem0003.clj
; Working off the PE explanation sheet. https://projecteuler.net/overview=003

(ns pe.clojure.implementation.problem0003)


(defn largestPrimeFactor [numberToFactor]
  (apply max (primeFactors numberToFactor)))


(defn primeFactors [numberToFactor]
  (loop [factorBy 2
         currentNumberToFactor numberToFactor
         primeFactors #{}]
    (cond
     (> factorBy currentNumberToFactor) primeFactors
     (= 0 (mod currentNumberToFactor factorBy)) (recur factorBy
                                                (/ currentNumberToFactor factorBy)
                                                (conj primeFactors factorBy))
     :else (recur (inc factorBy) currentNumberToFactor primeFactors))))

File: pe/nonvisualstudio/elixir/Implementation/Problem0003.ex
defmodule Problem0003 do

  def largestPrimeFactor(numberToFactor), do: primeFactors(numberToFactor) |> Enum.max

  def primeFactors(numberToFactor) do
    primeFactorsRecur(2, numberToFactor, HashSet.new)
  end

  defp primeFactorsRecur(factorBy, currentNumberToFactor, primeFactors) do
    cond do
      factorBy > currentNumberToFactor ->
        primeFactors
      isDivisibleByCurrentFactor(currentNumberToFactor, factorBy) ->
        primeFactorsRecur(factorBy, div(currentNumberToFactor, factorBy), primeFactors |> Set.put(factorBy))
      true ->
        primeFactorsRecur((factorBy + 1), currentNumberToFactor, primeFactors)
    end
  end

  defp isDivisibleByCurrentFactor(currentNumber, factorBy), do: rem(currentNumber, factorBy) == 0
end
